File: algorithms/2_medium/N/01524_number_of_sub_arrays_with_odd_sum/solution.c
#define MOD 1000000007

int numOfSubarrays(const int *restrict arr, const int len) {
    int32_t cnt = 0;
    int32_t odd = 0;
    int32_t even = 1;
    int32_t sum = 0;

    for (int i = 0; i < len; i += 1) {
        sum += arr[i];

        if ((sum & 1) == 1) {
            cnt = (cnt + even) % MOD;
            odd += 1;
        } else {
            cnt = (cnt + odd) % MOD;
            even += 1;
        }
    }

    return (int)cnt;
}

File: algorithms/2_medium/N/01524_number_of_sub_arrays_with_odd_sum/solution.rs
impl Solution {
    const MOD: i32 = 1_000_000_007;

    pub fn num_of_subarrays(arr: Vec<i32>) -> i32 {
        let mut cnt = 0_i32;
        let mut odd = 0_i32;
        let mut even = 1_i32;
        let mut sum = 0_i32;

        for a in arr {
            sum += a;

            if (sum & 1) == 1 {
                cnt = (cnt + even) % Self::MOD;
                odd += 1;
            } else {
                cnt = (cnt + odd) % Self::MOD;
                even += 1;
            }
        }

        cnt
    }
}
